src/events.rs

```diff
@@ -0,0 +1,319 @@
+    pending_watched_files: HashSet<PathBuf>,
+    event_stream: mpsc::UnboundedReceiver<Result<notify::Event, notify::Error>>,
+}
+
+impl Debug for MuxedEvents {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("MuxedEvents")
+            .field("watched_directories", &self.watched_directories)
+            .field("watched_files", &self.watched_files)
+            .field("pending_watched_files", &self.pending_watched_files)
+            .finish()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Failed to add path to watch")]
+    AddFailure,
+    #[error("Failed to remove path from watch")]
+    RemoveFailure,
+    #[error("Watch already exists for path")]
+    WatchExists,
+    #[error("Error receiving event: {0}")]
+    Event(#[from] std::io::Error),
+}
+
+impl MuxedEvents {
+    /// Constructs a new `MuxedEvents` instance.
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let inner: notify::RecommendedWatcher = notify::Watcher::new_immediate(move |res| {
+            // The only way `send` can fail is if the receiver is dropped,
+            // and `MuxedEvents` controls both. `unwrap` is not used,
+            // however, since `Drop` idiosyncrasies could otherwise result
+            // in a panic.
+            let _ = tx.send(res);
+        })
+        .expect("Failed to start watcher");
+
+        MuxedEvents {
+            inner,
+            watched_directories: HashMap::new(),
+            watched_files: HashSet::new(),
+            pending_watched_files: HashSet::new(),
+            event_stream: rx,
+        }
+    }
+
+    fn watch_exists(&self, path: impl AsRef<Path>) -> bool {
+        let path = path.as_ref();
+
+        // Make sure we aren't already watching the directory
+        self.watched_files.contains(&path.to_path_buf())
+            || self.pending_watched_files.contains(&path.to_path_buf())
+            || self.watched_directories.contains_key(&path.to_path_buf())
+    }
+
+    fn add_directory(&mut self, path: impl AsRef<Path> + Into<PathBuf>) -> Result<(), Error> {
+        let path_ref = path.as_ref();
+
+        // TODO: Check the count, but this is okay to call multiple times anyway
+        notify::Watcher::watch(
+            &mut self.inner,
+            &path_ref,
+            notify::RecursiveMode::NonRecursive,
+        )
+        .map_err(|_e| Error::AddFailure)?;
+
+        let count = self.watched_directories.entry(path.into()).or_insert(0);
+        *count += 1;
+
+        Ok(())
+    }
+
+    fn remove_directory(&mut self, path: impl AsRef<Path>) -> Result<(), Error> {
+        let path_ref = path.as_ref();
+
+        if let Some(count) = self.watched_directories.get(path_ref).copied() {
+            match count {
+                0 => unreachable!(), // watch is removed if count == 1
+                1 => {
+                    let _ = notify::Watcher::unwatch(&mut self.inner, &path_ref)
```

Comment:
Instead of completely dropping the error here, it might be better to still remove the path_buf from the list but to return an error. Although there's not much to do at that point so 🤷‍♂